src/main/java/spoon/support/reflect/reference/CtReferenceImpl.java

```diff
@@ -98,22 +98,92 @@ public boolean equals(Object o) {
 		}
 		return false;
 	}
-	private void checkIdentiferForJLSCorrectness(String simplename) {
-		/*
-		 * At the level of the Java Virtual Machine, every constructor written in the Java programming language (JLS §8.8)
-		 * appears as an instance initialization method that has the special name <init>.
-		 * This name is supplied by a compiler. Because the name is not a valid identifier,
-		 * it cannot be used directly in a program written in the Java programming language.
-		 */
-		//JDTTreeBuilderHelper.computeAnonymousName returns "$numbers$Name" so we have to skip them if they start with numbers
-		//allow empty identifier because they are sometimes used.
-		if (!simplename.matches("<.*>|\\d.*|^.{0}$")) {
-			//split at "<" and ">" because "Iterator<Cache.Entry<K,Store.ValueHolder<V>>>" submits setSimplename ("Cache.Entry<K")
-			String[] splittedSimplename = simplename.split("\\.|<|>");
-			if (checkAllParts(splittedSimplename)) {
-				throw new SpoonException("Not allowed javaletter or keyword in identifier found. See JLS for correct identifier. Identifier: " + simplename);
+
+	/*
+	 * This method validates the simplename.
+	 * spoon needs to allow more names that are allowed by the JLS, as
+	 * - array references have a name, e.g. int[], where [] would not be allowed normally
+	 * - ? is used as name for intersection types
+	 * - <init>, <clinit>, <nulltype> are used to represent initializers and the null reference
+	 * - anonymous/local classes start with numbers in spoon
+	 * - simple names of packages are just their names, but they may contain '.'
+	 * - the name can contain generics, e.g. List<String>[]
+	 */
+	private void checkIdentifierForJLSCorrectness(String simplename) {
+		if (isSpecialType(simplename)) {
+			return;
+		}
+		if (!checkAll(simplename)) {
+			throw new SpoonException("Not allowed javaletter or keyword in identifier found. See JLS for correct identifier. Identifier: " + simplename);
+		}
+	}
+
+	/*
+	 * returns true if the name is valid.
+	 * this splits up the string into parts that need to be JLS compliant.
+	 */
+	private boolean checkAll(String name) {
+		int i = 0;
+		// leading digits come from anonymous/local classes. Skip them
+		while (i < name.length() && Character.isDigit(name.charAt(i))) {
+			i++;
+		}
+		int start = i; // used to mark the beginning of a part
+		char expectNext = 0; // 0 = do not expect anything
```

Comment:
Avoid unnamed magic values :)

```suggestion
		final char anything = 0;
		char expectNext = anything;
```